weave/flow/eval2.py

```diff
@@ -0,0 +1,200 @@
+
+import weave
+from weave.flow.eval import Evaluation, default_evaluation_display_name
+from weave.flow.model import Model
+from weave.flow.scorer import Scorer
+from weave.trace.context import call_context
+from weave.trace.context.weave_client_context import require_weave_client
+from weave.trace.weave_client import Call
+
+ID = str
+
+NOT_DEFINED = "Not defined for custom scoring"
+
+# Context variable to store the current output safely between threads.  This also
+# ensures that only 1 version of the predict method is saved because the code
+# contents are always the same.
+current_output: ContextVar[Any] = ContextVar("current_output", default=None)
+current_score: ContextVar[float | None] = ContextVar("current_score", default=None)
+current_summary: ContextVar[dict | None] = ContextVar("current_summary", default=None)
+
+
+@contextmanager
+def set_current_output(output: Any) -> Iterator[None]:
+    """Set the current output in a thread-safe way using context variables."""
+    token = current_output.set(output)
+    try:
+        yield
+    finally:
+        current_output.reset(token)
+
+
+@contextmanager
+def set_current_score(score: float) -> Iterator[None]:
+    token = current_score.set(score)
+    try:
+        yield
+    finally:
+        current_score.reset(token)
+
+
+@contextmanager
+def set_current_summary(summary: dict) -> Iterator[None]:
+    token = current_summary.set(summary)
+    try:
+        yield
+    finally:
+        current_summary.reset(token)
+
+
+class BetaScoreLogger(BaseModel):
+    prediction_id: ID
+    score: float
+    metadata: dict | None
+
+
+class BetaPredictionLogger(BaseModel):
+    model_id: ID
+    inputs: dict
+    output: Any
+    hack_reference_to_predict_and_score: Call
+    hack_reference_to_evaluate: Call
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    async def log_score(
+        self, scorer_name: str, score: float, metadata: dict | None = None
+    ) -> BetaScoreLogger:
+        # Define the scorer class dynamically
+        cls_name = scorer_name.replace(".", "_")
+        DynamicScorer = type(cls_name, (Scorer,), {})
+        scorer_instance = DynamicScorer()
+
+        @weave.op(name=scorer_name)
+        def score_method(self: Scorer, *, output: Any, **inputs: Any) -> float:
+            # return score
+            # TODO: can't use score here because it will cause version mismatc
+            return cast(float, current_score.get())
+
+        scorer_instance.__dict__["score"] = MethodType(score_method, scorer_instance)
+
+        # attach the score feedback to the predict call
+        with call_context.set_call_stack(
+            [self.hack_reference_to_evaluate, self.hack_reference_to_predict_and_score]
+        ):
+            await self.hack_reference_to_predict_and_score.apply_scorer(scorer_instance)
+
+        return BetaScoreLogger(prediction_id="123", score=score, metadata=metadata)
+
+
+class BetaEvaluationLogger(BaseModel):
+    model_id: ID = ""
```

Comment:
I am not seeing `model_id` actually consumed anywhere. I think that `model_id` needs to be somewhat a property of the created model itself. I also think a dictionary would be good here as it can uniquely identify the model based on properties.

Reply:
I don't understand the dictionary comment, can you expand?